tests/APIv0/Controllers/ModerationControllerTest.php

```diff
@@ -0,0 +1,484 @@
+     * where destination category already has LastDiscussionID fresher than we move in.
+     *
+     * @large
+     *
+     * @depends testInitDB
+     * @depends testMoveNewDiscussionsEmptyCategories
+     * @depends testMoveExistingDiscussionToEmptyCategory
+     */
+    public function testMoveExistingDiscussionToNotEmptyCategory() {
+        $destCatKey = 'cat2_1';
+        $this->moveDiscussion(self::$discussions['d1_case1'], 'cat3_1_1', $destCatKey, self::$discussions['d1_case4']);
+        $this->moveDiscussion(self::$discussions['d1_case2'], 'cat3_1_1', $destCatKey, self::$discussions['d1_case4']);
+        $this->recheckCategories();
+    }
+    /**
+     * Use case #4:
+     * Move most recent discussion from one category to another
+     *
+     * @large
+     *
+     * @depends testInitDB
+     * @depends testMoveNewDiscussionsEmptyCategories
+     * @depends testMoveExistingDiscussionToEmptyCategory
+     * @depends testMoveExistingDiscussionToNotEmptyCategory
+     */
+    public function testMoveMostRecentDiscussionToNotEmptyCategory() {
+        $this->moveDiscussion(self::$discussions['d1_case6'], 'cat2_1_1_1_1_1_1', 'cat2_1_1');
+        $this->moveDiscussion(self::$discussions['d1_case5'], 'cat2_1_1_1_1_1', 'cat3_1_1');
+        $this->moveDiscussion(self::$discussions['d1_case5'], 'cat3_1_1', 'cat2_1', self::$discussions['d1_case4']);
+        $this->recheckCategories();
+    }
+    /**
+     * Use case #5:
+     * New comment add to existing discussion and move that discussion after
+     *
+     * @large
+     *
+     * @depends testInitDB
+     * @depends testMoveNewDiscussionsEmptyCategories
+     * @depends testMoveExistingDiscussionToEmptyCategory
+     * @depends testMoveExistingDiscussionToNotEmptyCategory
+     * @depends testMoveMostRecentDiscussionToNotEmptyCategory
+     */
+    public function testAddCommentMoveDiscussion() {
+        $comment = $this->addComment([
+            'DiscussionID' => self::$discussions['d1_case1']['DiscussionID'],
+            'Body' => 'Moderation controller test.'
+        ], 'd1_case1', 'cat2_1');
+        $this->recheckCategories();
+        $this->moveDiscussion(self::$discussions['d1_case1'], 'cat2_1', 'cat3_1_1', self::$discussions['d1_case6']);
+        $this->recheckCategories();
+    }
+    /**
+     * Use case #6:
+     * Create new comment to existing discussion and move it to some destCategory
+     * After that move we expect it to have most recent fields updated
+     * Then move some previous discussion into the same destCategory
+     * We expect detCategory recent field to not be updated
+     *
+     * @large
+     *
+     * @depends testInitDB
+     * @depends testMoveNewDiscussionsEmptyCategories
+     * @depends testMoveExistingDiscussionToEmptyCategory
+     * @depends testMoveExistingDiscussionToNotEmptyCategory
+     * @depends testMoveMostRecentDiscussionToNotEmptyCategory
+     * @depends testAddCommentMoveDiscussion
```

Comment:
I'm concerned with how tightly these tests seem to be coupled. It looks like you're relying on data being set in one test in order to run another. There's a heavy dependence on the "depends" tag, which would typically be used to skip a test because functionality it relies on failed its tests (e.g. "don't test revoking access tokens, if we can't issue them"), but the usage here appears to be to ensure the referenced category and discussion data has been set from other tests.

A better way to handle your site data might be to create your category structure in a `setupBeforeClass` (or reset it in a `setup` method) and add your relevant post data only in each test. I believe this would allow you to run specific tests without the current "dependencies" that require almost all tests to be run, in some cases.

Reply:
As usually, I can agree on theory. But in reality, if we set "good answer" (expected) against "unit test" that should work. But in reality we have integration tests. And all these scenarios appeared only because I did run them in different order and that helped me to discover more and more weird behavior  and weird values. Right now it "seems" or "looks like" we know "expected" good values. But in fact we don't. As Lincoln mentioned during one of our discussion, probably best approach is to do step back and to set clear technical requirements document and then we can set up clear unit tests even before we start an implementation. But in our particular case we have code running without any description of what that code should do and what it shouldn't.

If you want, I can split this big "integration" test into pieces (I can kind do snapshot after each step and store intermediate results), but I have strong opinion that it is wrong approach which will lead to tests which can detect much less issues.

Reply:
Another option: we can just transform several tests into one big test with consecutive actions and assertions

Reply:
If you want to make a larger test, that's fine, but my basic point is that I should have the option of running one test or every test in the "test case". Each test should be verifying specific behavior, so it shouldn't require an unnecessarily long list of dependencies. Here, attempting to run `testCommentDestinationLastCommentFresher` shouldn't require half a dozen other tests to run that don't functionally preclude this one test from running. In this particular case, where we're testing to make sure the last comment was updated when a discussion is moved, it *might* make sense to have `testAddCommentMoveDiscussion` as a dependency, but that's about it. Even then, it's only a dependency because if we aren't moving discussions properly, there's no point in checking the comments. The contents of `testCommentDestinationLastCommentFresher` shouldn't care about what was written to the database in other tests. It's self-contained. If it's needs data, it generates it, unless that data was made available by one of the setup methods.

Reply:
I am 100% sure the problem is the language. I just see things and can not express myself.

>  there's no point in checking the comments.
That would be true if we had an application created with respect of some technical requirements. But we don't. My way to test helped me to discover weird side effects in data changes. That is something you did care from the beginning. You wanted this test to test flow to assure that next implementation will not break the current data changes even if those are weird and unexpected. 

If I would follow your logic I would never probably find this bug:
https://github.com/vanilla/vanilla/issues/7897

Because according to general logic that should never exist. So why should we test that case?
But I prefer to "play" with data and top "play" with application. I do use them as "black box" and send same requests or very similar just to confirm that  result remains the same. I will work to transform this sequence of tests to set of independent tests. But I disagree that it is something good. The whole story of why that became necessary to be made is against this approach.

Reply:
I said "there's no point in checking the comments", **if the `testAddCommentMoveDiscussion` failed**. I did not say you should not test for that behavior, at all. I was illustrating a *good* usage of test dependencies. With the current test setup, you could add `testAddCommentMoveDiscussion` (and only `testAddCommentMoveDiscussion`) as a dependency to `testCommentDestinationLastCommentFresher`. If we were failing to move a discussion and it's comments, it would make sense to skip testing something that depends on discussions/comments moving. Dependencies shouldn't be used as a way to order your tests so you'll have test data by the time you execute the target.

I'm all for you testing whatever scenarios and data you can think of. I just want them to be structured in a sane way.